src/firewall.ts

```diff
@@ -460,147 +469,300 @@ class WirefilterFirewallRule implements FirewallRule {
     }
 
     match(req: Request): boolean {
-        const wirefilter = this.wirefilter;
-        const exec_ctx = wirefilter.wirefilter_create_execution_context(this.scheme);
+        const exec_ctx = ExecutionContext.buildFromRequest(this.wirefilter, this.scheme, this.lists, req);
+
+        try {
+            return this.matchUsingContext(exec_ctx);
+        } finally {
+            exec_ctx.free();
+        }
+    }
+
+    matchUsingContext(execCtx: ExecutionContext): boolean {
+        const matchResult = this.wirefilter.wirefilter_match(this.filter, execCtx.internalPointer);
+        if (matchResult.ok.success != 1) {
+            throw Error(`Filter can't be matched: ${matchResult.err.msg.data}`);
+        }
+        return matchResult.ok.value;
+    }
+}
+
+
+type Phase = 'http_ratelimit' | 'http_request_sbfm' | 'http_request_firewall_managed'
+type Product = 'zoneLockdown' | 'uaBlock' | 'bic' | 'hot' | 'securityLevel' | 'rateLimit' | 'waf'
+
+type FirewallAction =
+    | {
+        type: 'skip'
+        ruleset?: 'current'
+        phases?: Phase[]
+        products?: Product[]
+    }
+    | {
+        type: 'block' | 'challenge' | 'managed_challenge' | 'js_challenge' | 'log'
+    }
+
+type Rule = {
+    id: string
+    expression: WirefilterFirewallRule
+    action: FirewallAction
+}
+
+type RulesetMatchResult = {
+    loggedRules: string[]
+    skippedPhases: Set<Phase>
+    skippedProducts: Set<Product>
+    outcome:
+    | {
+        action: | 'allow'
```

Comment:
think the default here should be undefined or `no_match` rather than allow.

Avoids the confusion with skip or the old allow action

Reply:
Makes sense, will update.